### ISSUE TITLE
Add support for local LLMs via ollama openai compatible api.

### ISSUE THREAD

**Issue by palindsay** [copilot-emacs/copilot.el#302]
It would nice to be able to use ollama with local LLMs versus github co-pilot.

**Comment by LemonBreezes** [comment]
Yeah. This would be great. Ollama does provide a completions API endpoint so there should be a way.

**Comment by pakelley** [comment]
Looks like [ollama-copilot](https://github.com/bernardo-bruning/ollama-copilot) has [instructions for setting up neovim](https://github.com/bernardo-bruning/ollama-copilot?tab=readme-ov-file#neovim) using copilot.vim.
Seeing as copilot.el is using copilot.vim, seems possible we could use this and set `copilot-network-proxy` accordingly? Though from a quick glance at the copilot.el code I don't see anything about the strict ssl setting, so setting that might be an issue.

Hoping to experiment with this myself soon, but in case someone beats me to it 🙂

**Comment by kokangit** [comment]
> Looks like [ollama-copilot](https://github.com/bernardo-bruning/ollama-copilot) has [instructions for setting up neovim](https://github.com/bernardo-bruning/ollama-copilot?tab=readme-ov-file#neovim) using copilot.vim. Seeing as copilot.el is using copilot.vim, seems possible we could use this and set `copilot-network-proxy` accordingly? Though from a quick glance at the copilot.el code I don't see anything about the strict ssl setting, so setting that might be an issue.
> 
> Hoping to experiment with this myself soon, but in case someone beats me to it 🙂

It actually works!
Regarding "strict ssl": have a look at the variable `copilot-network-proxy` in copilot.el. You should use `:rejectUnauthorized :json-false`.

**Comment by knilink** [comment]
for anyone who is interested, i also made a tool to mod the language server script into local copilot without using a proxy, or even without signing up a copilot account https://github.com/knilink/tamperpilot

**Comment by mikej96** [comment]
@kokangit  so you got this to work? Does it work well in practice?

**Comment by ekwessel** [comment]
@kokangit I'd also like to follow up and ask if you (or anyone) has gotten this to work. I can get the copilot.el extension to talk to ollama-copilot somewhat, in the sense that when I type in the editor requests are sent to ollama according to its output. However, I can't seem to get any responses back, so I'm not seeing any completions. I'm slightly off the beaten path here, and there is literally no documentation anywhere besides this thread on doing this (fairly simple) thing, so if you got it working an example would be highly appreciated!

**Comment by ProjectMoon** [comment]
@ekwessel I got it to work, but it's a very fragile setup. An interesting alternative you might consider until/if support is added here would be Starhugger: https://github.com/daanturo/starhugger.el

**Comment by ekwessel** [comment]
I'll certainly look into trying that next! However, since I'm *so close* to getting this working, I'm curious about what I must be missing. I installed ollama-copilot the usual way. I've launched it manually, then started doom emacs. In my doom config.el I've added (after copilot.el's `!use-package` block):
```
(after! copilot
  (setq copilot-network-proxy '(:host "127.0.0.1" :port 11435 :rejectUnauthorized :json-false)))
```
However, when I type, I get no completions. According to the ollama-copilot output, *something* is happening, I get output like this:
```
2025/01/15 09:35:26 request: GET /copilot_internal/v2/token
2025/01/15 09:35:26 response: GET /copilot_internal/v2/token 200 314.834µs
2025/01/15 09:35:27 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:35:27 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:35:28 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:35:28 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:35:29 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:35:30 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:35:32 request: POST /v1/engines/copilot-codex/completions
2025/01/15 09:37:31 failed to transfer: writeto tcp 10.142.206.180:50819->13.107.5.93:443: readfrom tcp 127.0.0.1:11435->127.0.0.1:50816: read tcp 10.142.206.180:50819->13.107.5.93:443: read: connection reset by peer
```
But I don't really understand what is going wrong or what I could do about it.

**Comment by gamboz** [comment]
Hi @ekwessel I've got completions from ollama + ollama-copilot when I use `:port 11434` (I've proposed my config to ollama-copilot [here](https://github.com/bernardo-bruning/ollama-copilot/pull/24/files) ), but I'm not sure why it works 🤕 

@ProjectMoon thanks for the pointer to starhugger 🙂
